packages/x-data-grid-pro/src/hooks/features/infiniteLoader/useGridInfiniteLoader.tsx

```diff
@@ -0,0 +1,144 @@
+import * as React from 'react';
+import {
+  useGridSelector,
+  useGridApiOptionHandler,
+  gridVisibleColumnDefinitionsSelector,
+  useGridApiMethod,
+  gridDimensionsSelector,
+} from '@mui/x-data-grid';
+import { useGridVisibleRows } from '@mui/x-data-grid/internals';
+import useEventCallback from '@mui/utils/useEventCallback';
+import { styled } from '@mui/system';
+import type { GridInfiniteLoaderPrivateApi } from '@mui/x-data-grid/models/api/gridInfiniteLoaderApi';
+import { GridRowScrollEndParams } from '../../../models';
+import { GridPrivateApiPro } from '../../../models/gridApiPro';
+import { DataGridProProcessedProps } from '../../../models/dataGridProProps';
+
+const InfiniteLoadingTriggerElement = styled('div')({
+  position: 'sticky',
+  left: 0,
+  width: 0,
+  height: 0,
+});
+
+/**
+ * @requires useGridColumns (state)
+ * @requires useGridDimensions (method) - can be after
+ * @requires useGridScroll (method
+ */
+export const useGridInfiniteLoader = (
+  apiRef: React.MutableRefObject<GridPrivateApiPro>,
+  props: Pick<
+    DataGridProProcessedProps,
+    'onRowsScrollEnd' | 'pagination' | 'paginationMode' | 'rowsLoadingMode' | 'scrollEndThreshold'
+  >,
+): void => {
+  const visibleColumns = useGridSelector(apiRef, gridVisibleColumnDefinitionsSelector);
+  const currentPage = useGridVisibleRows(apiRef, props);
+  const observer = React.useRef<IntersectionObserver>();
+  const triggerElement = React.useRef<HTMLElement | null>(null);
+
+  const isEnabled = props.rowsLoadingMode === 'client' && !!props.onRowsScrollEnd;
+
+  const handleLoadMoreRows = useEventCallback(([entry]: IntersectionObserverEntry[]) => {
+    const currentRatio = entry.intersectionRatio;
+    const isIntersecting = entry.isIntersecting;
+
+    if (isIntersecting && currentRatio === 1) {
+      const viewportPageSize = apiRef.current.getViewportPageSize();
+      const rowScrollEndParams: GridRowScrollEndParams = {
+        visibleColumns,
+        viewportPageSize,
+        visibleRowsCount: currentPage.rows.length,
+      };
+      apiRef.current.publishEvent('rowsScrollEnd', rowScrollEndParams);
+      if (observer.current) {
+        if (triggerElement.current) {
+          observer.current?.unobserve(triggerElement.current);
+          // do not observe this node anymore
+          triggerElement.current = null;
+        }
+      }
+    }
+  });
+
+  const virtualScroller = apiRef.current.virtualScrollerRef.current;
+  const dimensions = useGridSelector(apiRef, gridDimensionsSelector);
+
+  React.useEffect(() => {
+    if (!isEnabled) {
+      return;
+    }
+    if (!virtualScroller) {
+      return;
+    }
+    if (triggerElement.current) {
+      observer.current?.unobserve(triggerElement.current);
+    }
```

Comment:
Not sure if it applies but there is also `observer.current?.disconnect()`, which can help avoid storing the observed element ref sometimes.

Reply:
Good point! I used `disconnect` instead of unobserve.

I still need to keep track of the trigger element to observe it after recreating `IntersectionObserver` (e.g. when `marginBottom` changes): https://github.com/mui/mui-x/pull/8672/files#diff-626d69f7d7c9aebfab3fdc6388cd0f88a38b6d9e51c567e76f63a25d63895b10R81-R83